### PR TITLE
RFC note major revision

### PR DESCRIPTION
**Pull request by jvonau**
Bump iiab_revision as the base point for a tagged pre-release (RC). 
It's easier to find than using the git hash to denote where the base code is at currently.

### CODE REVIEW

**Comment by holta** [comment]
I'd favor making the Date-of-Cloning-from-repo (e.g. YYYY-MM-DD HH:MM:SS TZ ?) more prominent in places where that should be made much more visible:

- top of [iiab-diagnostics](https://github.com/iiab/iiab/blob/master/scripts/iiab-diagnostics.README.md) output?
- other places?
- iiab-install-date command?

**Comment by jvonau** [comment]
> I'd favor making the Date-of-Cloning-from-repo (e.g. YYYY-MM-DD HH:MM:SS TZ ?) more prominent in places where that should be made much more visible:

Currently that information is already recorded in the iiab.ini file sections [version] and [runtime] when iiab-diagnostics is summited. Intended to be used as a marker for a point in time in the release cycle, there has been no tagged RC for a while now. Doesn't have to be done today, but this has become more of a rolling release, this is an easy marker for progress.

**Comment by tim-moody** [comment]
An install today is more like a nightly than a release. The 'major revision' seems more like a 'minor revision', e.g. 7.2.1 vs 7.2.0. but specifies a release more than a nightly.

**Comment by holta** [comment]
> An install today is more like a nightly than a release.

Similar.  But our CI/CD/QA is substantially stronger than many "nightly build" projects, which very regularly post broken builds.

Which is the reason many organizations strongly favor our latest IIAB install (master branch and associated [1-line installer](http://download.iiab.io)) — as a result of our quite strong Rolling QA community ethic.

**Comment by jvonau** [comment]
> An install today is more like a nightly than a release. The 'major revision' seems more like a 'minor revision', e.g. 7.2.1 vs 7.2.0. but specifies a release more than a nightly.

Sorry the commit message was misleading, but that is exactly what I was trying to imply, a known working combination based on Rolling QA that the install is working correctly.

**Comment by tim-moody** [comment]
What does iiab_base_ver mean, the latest release? It isn't.
If iiab_revision is a minor version, why not just tag it onto iiab_base_ver?
Does it really make sense to have all these numbers that don't line up with the release cycle? I'm not clear what they are used for.
I don't think that a commit necessarily constitutes 
>a known working combination based on Rolling QA

**Comment by holta** [comment]
There's a lot of positive (and certainly some negative!) that comes out of "mandatory" monthly-like releases, similar to many browsers.  Some very successful projects have a mandatory Beta release exactly every 2 weeks.  These are very worthwhile models, when they foster genuine collaboration, as they can do.

Of course many of those initiatives have Highly-Capitalized marketing, documentation and QA teams containing many hundreds of people to make Release Splash very real — so the parallels aren't perfect.

A different kind of outreach is necessary for projects / products like IIAB.  I'm very hopeful that very new kinds of outreach are in fact possible over the coming year.

**Comment by jvonau** [comment]
> What does iiab_base_ver mean, the latest release? It isn't.

I agree, this 7.2 release cycle was started when #2435 was pushed, and  iiab_base_ver should be bumped at some point again right? iiab_base_ver is intended to note the release but nothing has released yet, just revised.

> If iiab_revision is a minor version, why not just tag it onto iiab_base_ver?

Finer grained notes, iiab_revision would be reset to 0 once iiab_base_ver is no longer referring to this current cycle. Sure I agree you can put all that in one variable, bumping either variable has upkeep as the downside, upside is better versioning of the code base.

> Does it really make sense to have all these numbers that don't line up with the release cycle? I'm not clear what they are used for.

7.2.0 yesterday -> 7.2.1 today -> 7.2.2 tomorrow's new feature -> 7.2.3 next week's new feature -> 8.0.0 when enough is enough. Just a little more effort in using what already is present but not fully utilized.

**Comment by holta** [comment]
What about an auto-generated human-readable version number?

Auto-mapped from the latest merged PR or auto-mapped from the latest merged commit?  Or based on their GMT date or similar?

Some folks really like counting numbers (whether starting at zero or starting at 1) instead of dates, similar to Sugar Activity version numbers, that (in theory!) avoid all decimal points.  But either has benefits (date or counting numbers) if it's very short, distinctive and human-readable — and beneficial to community conversation.

**Comment by jvonau** [comment]
> There's a lot of positive (and certainly some negative!) that comes out of "mandatory" monthly-like releases, similar to many browsers. Some very successful projects have a mandatory Beta release exactly every 2 weeks. These are very worthwhile models, when they foster genuine collaboration, as they can do.

I was thinking more of using a git tag as a trigger for a GH-action to build some images but I'm second guessing myself.

**Comment by tim-moody** [comment]
What is the process by which 7.2.0 becomes 7.2.1, every merged commit, community decision? Wouldn't it be simpler just to get back to releases, even minor releases? Is there existing code that uses these variables other than just recording them?

There's actually a GH mechanism for generating these numbers. It's called a release.

**Comment by jvonau** [comment]
> What is the process by which 7.2.0 becomes 7.2.1, every merged commit, community decision? 

I could see community decision, or perhaps a role's revision as in #2533's [upgrade_roles](https://github.com/iiab/iiab/pull/2533/files#diff-8abc169bf41dd311f233f504793fa3b07ab67ac7854d2dc5cf487c29ff1e9b70) which holds the role to revision relationship with [upgrade_roles.txt](https://github.com/iiab/iiab/pull/2533/files#diff8abc169bf41dd311f233f504793fa3b07ab67ac7854d2dc5cf487c29ff1e9b70) providing expanded information on the change to [iiab_revision](https://github.com/iiab/iiab/pull/2533/files#diff-238bd4eb09be3ed2de08302dcea3da77b274c8f4f3181c2f950c96dbcd6902b7)

Should a standard practice of bumping iiab_revision as part of pushing a PR the need for the external list I used above goes away.
   
> Wouldn't it be simpler just to get back to releases, even minor releases? 

I would prefer that but there seems to be an effort to drag out releases.
> Is there existing code that uses these variables other than just recording them?

I've cut some code for testing auto-upgrading of some roles, used with 2533 above.

> There's actually a GH mechanism for generating these numbers. It's called a release.

Sure as David from the past once said "Release early, release often" add to that "but you have to release".

**Comment by tim-moody** [comment]
>I would prefer that but there seems to be an effort to drag out releases

perhaps it's time to get that back on the agenda rather than creating workarounds

**Comment by jvonau** [comment]
> > I would prefer that but there seems to be an effort to drag out releases
> 
> perhaps it's time to get that back on the agenda rather than creating workarounds

It's not really a workaround rather it is a sound refinement in the release process that has been ignored, is it just too hard to bump the iiab_revision (or iiab_release should iiab_revisions go away) when a role's base point has changed or even by agreement on when to bump? Now iiab_revisions are not being maintained by choice, but craft up a GH-action or find one that fits then the 'by hand' maintenance goes away.

> What about an auto-generated human-readable version number?
> 
> Auto-mapped from the latest merged PR or auto-mapped from the latest merged commit? Or based on their GMT date or similar?
> 
> Some folks really like counting numbers (whether starting at zero or starting at 1) instead of dates, similar to Sugar Activity version numbers, that (in theory!) avoid all decimal points. But either has benefits (date or counting numbers) if it's very short, distinctive and human-readable — and beneficial to community conversation.

Sure I look forward to your workflow proposal.

**Comment by holta** [comment]
Every commit to master branch(es) is a release, in our open source model.  That can be exposed with human-readable-by-everyone counting numbers (or easy-to-read dates) if we choose such a scheme — now that Rolling QA (CD = Continuous Deployment, as truly exhausting as it can sometimes be) has been proven to work — ramping up participatory & timely community quality.

These probably should be auto-exposed (auto-published) if an easy way can be found, for the [GitHub Actions](https://github.com/features/actions) Jerry outlined, and similar to Tim's "nightly build" comparison — allowing different sub-communities to converge on any historic release that they like, if they have an easier way [de facto version number/name] to talk about it (without living in the past, just for the sake of ritualized vintage nostalgia!)

Quite separately, there is an absolutely **huge** amount of documentation & outreach hard work that goes into creating any community "marketing" and social/media campaigning for a very public release, and the Media Splash to make it legit effective.  The 2 above things are completely and entirely different.  Conflating them can of course be tempting.  But it is also a mistake.

No less when it accidentally demotivates (unintentionally, and unquestionably) those artistic/design/humanitarian volunteers whose natural skillsets will help us with that upcoming Vital & Very Hard Work.  Public advocacy, campaigning, communications, private persuasion, etc (i.e. inspiring the globe to use radically empowering learning products like IIAB, long overdue for sure!)

Not at all easy, and if it were easy it would have been done long ago.  Thankfully it's a challenge that the world is ready for.  As (IIAB & our larger ecosystem) all together have a LOT to offer educators of almost all kinds, that will allow journalists and other activist/NGO/government opinion makers and other adoption channels to take this VERY seriously in 2021 and 2022.

No there won't be fireworks every day.

_But certainly they are coming._

**Comment by holta** [comment]
Aside: I personally (unfortunately) don't know anything about GitHub Actions (https://github.com/features/actions) but if anybody discovers that this is extremely helpful (for anything we need & do?) let's investigate!

**Comment by jvonau** [comment]
> Every commit to master branch(es) is a release, in our open source model. 

I'm not trying to be antagonistic, I would not call each commit a release but a series of commits in a PR for new functionality for the latest new role, that could be considered a release as there is a new feature. Documentation is not fun at all, bumping a upstream release number to use in a role is maintenance to keep up with the latest, looking out for gotcha burns from upstream updates, that is a never ending battle, but needs to be done.

>That can be exposed with human-readable-by-everyone counting numbers (or easy-to-read dates) if we choose such a scheme — now that Rolling QA (CD = Continuous Deployment, as truly exhausting as it can sometimes be) has been proven to work — ramping up participatory & timely community quality.

That might be true of the overall deliverable, but each sub-component in the pipeline more or less works on a Continuous Delivery(debs, pip installs, upstream git repos) model. It is (was?) common practice to provide some sort of [change.log](https://keepachangelog.com/en/1.0.0/) that tracks the progression of the code base as part of any Continuous Delivery model's version tracking. Yup I'm old school, like a summary in one place. Yes it is a big job, but when incremental changes can be tracked via a change.log in the workflow, each change is summarized as you go. The above upgrade_roles.txt link was an opening shot at some sort of change.log of what was changed and when, have to start somewhere, something is better than nothing. Reproducible results is the end goal right?

> 
> These probably should be auto-exposed (auto-published) if an easy way can be found, for the [GitHub Actions](https://github.com/features/actions) Jerry outlined, and similar to Tim's "nightly build" comparison — allowing different sub-communities to converge on any historic release that they like, if they have an easier way to talk about it (without living in the past, just for the sake of ritualized vintage nostalgia!)

Don't insinuate I live in the past, far from it, I'm not a beginner at Continuous Integration just rusty. Again I'm open to review whatever you or whomever you recruit, cooks up. 
> 
> Quite separately, there is an absolutely **huge** amount of documentation & outreach hard work that goes into creating any community "marketing" and social/media campaigning for a very public release, and the Media Splash to make it legit effective. The 2 above things are completely and entirely different. Conflating them can of course be tempting. But it is also a mistake.

Please don't stray into "marketing" I have a few choice thoughts on that subject you may not want to hear. The subject is software engineering, lets stay focused on the subject matter please.
> 
> No less when it accidentally demotivates (unintentionally, and unquestionably) those artistic/design/humanitarian volunteers whose natural skillsets will help us with that upcoming Vital & Very Hard Work. Public advocacy, campaigning, communications, private persuasion, etc (i.e. inspiring the globe to use radically empowering learning products like IIAB, long overdue for sure!) 

Using 'demotivates' in this context I understand, why I still debug issues for this project I'm not sure.
>Not at all easy, and if it were easy it would have been done long ago. Thankfully it's a challenge that the world is ready for. As (IIAB & our larger ecosystem) we have a LOT to offer educators of almost all kinds, that will allow journalists and other activist/NGO/government opinion makers and other adoption channels to take this VERY seriously in 2021 and 2022.

What is the phase "larger ecosystem" meant to infer? Receiving an issue about an installed application and forwarding the issue upstream? If you have a Continuous Deployment model the docs should point to changes to the deliverable at any given point in time. Dig though the git history of the repo used as part of the deliverable? It's not iiab per say, to analyze any issue reported with iiab-diagnostic you are provided with the runtime git hash,branch,date, looking up that git hash is a pain but you must to get any of the versions of the apps that are installed, because the versions are not recorded anywhere in the iiab.ini file but exists only in the roles' default/main.yml file. This information is nice to know at a glance but must be deduced, in place of user reports using 7.2.11 but 2.7.14 is out now, check the change.log, ok z,x,y changed, reported issue is similar to x, hum that might be it, hum, try reinstalling that role or just start over but the content is already loaded, never easy.
 
> Aside: I personally (unfortunately) don't know anything about GitHub Actions (https://github.com/features/actions) but if anybody discovers that this is extremely helpful (for anything we need & do?) let's investigate!

These newer GitHub Actions are starting to replacing Travis, perhaps you can approach your "larger ecosystem of fellow developers" that are producing standalone Rpi images for the same applications that IIAB installs for help or cruse their .github/workflows directory in their repos for inspiration on how to setup a pipeline.

**Comment by holta** [comment]
> git hash is a pain but you must to get any of the versions of the apps that are installed, because the versions are not recorded anywhere in the iiab.ini file but exists only in the roles' default/main.yml file

It's worse than that.  Many app/service/role version numbers are not contained in their defaults/main.yml

This might possibly be an important improvement that [iiab-diagnostics](https://github.com/iiab/iiab/blob/master/scripts/iiab-diagnostics.README.md) needs.

The info can be gleaned from each app/service if someone has Command-Line access, but those tricks are highly non-obvious, often a whole lot trickier than...

`COMMAND --version`

e.g. the ugly notes below contain most all of the tricks needed to identify IIAB app/service versions.  But still we should move towards better way(s) of surfacing these details:

http://FAQ.IIAB.IO > "28 Can I upgrade or reinstall server apps?"

**Comment by jvonau** [comment]
> > git hash is a pain but you must to get any of the versions of the apps that are installed, because the versions are not recorded anywhere in the iiab.ini file but exists only in the roles' default/main.yml file
> 
> It's worse than that. Many app/service/role version numbers are not contained in their defaults/main.yml
>

These should be more or less packages supplied as deb files by the system package manager. 
 
> This might possibly be an important improvement that [iiab-diagnostics](https://github.com/iiab/iiab/blob/master/scripts/iiab-diagnostics.README.md) needs.
>
Couldn't hurt, better information can set the stage for a faster or better solution.  
 
> The info can be gleaned from each app/service if someone has Command-Line access, but those tricks are highly non-obvious, often a whole lot trickier than...
> 
> `COMMAND --version`
> 

Those would be non apt installed apps, those are a bit of a pain and are usually optional.

> e.g. the ugly notes below contain most all of the tricks needed to identify IIAB app/service versions. But still we should move towards better way(s) of surfacing these details:
> 
> http://FAQ.IIAB.IO > "28 Can I upgrade or reinstall server apps?"

One of the bits of information is the local_vars.yml file, this information is exceptionally hard to parse with the notes and settings intermingled, the information is there but picking out the setting is not that easy, at least for me. The local_vars.yml is pretty long winded and has multiple sources, having one place for all the notes and histories kind of makes sense to have a local_vars.Readme with all that information and pear down what the 3 sisters offer to install like @tim-moody's [medical.yml](https://github.com/iiab/iiab/blob/master/vars/local_vars_medical.yml) currently is setup, clear and easy to read.

**Comment by jvonau** [comment]
waste of my time
